File: _sqlalchemy/loaddata.py
#
# Copyright (c) 2019 MagicStack Inc.
# All rights reserved.
#
# See LICENSE for details.
##


import argparse
import collections
import datetime
import json
import os
import progress.bar

import sqlalchemy as sa
import sqlalchemy.orm as orm

import _sqlalchemy.models as m


def bar(label, total):
    return progress.bar.Bar(label[:32].ljust(32), max=total)


def bulk_insert(db, label, data, into):
    label = f"Creating {len(data)} {label}"
    pbar = bar(label, len(data))

    while data:
        chunk = data[:1000]
        data = data[1000:]
        db.execute(sa.insert(into), chunk)
        db.commit()
        pbar.next(len(chunk))
    pbar.finish()


def reset_sequence(db, tablename):
    tab = sa.table(tablename, sa.column("id"))

    db.execute(
        sa.select(
            sa.func.setval(
                f"{tablename}_id_seq",
                sa.select(tab.c.id)
                .order_by(tab.c.id.desc())
                .limit(1)
                .scalar_subquery(),
            )
        )
    )


def load_data(filename, engine):
    session_factory = orm.sessionmaker(bind=engine)
    Session = orm.scoped_session(session_factory)

    with Session() as db:

        # first clear all the existing data
        print(f"purging existing data...")

        db.execute(sa.delete(m.Directors))
        db.execute(sa.delete(m.Cast))
        db.execute(sa.delete(m.Review))
        db.execute(sa.delete(m.Movie))
        db.execute(sa.delete(m.Person))
        db.execute(sa.delete(m.User))
        db.commit()

    # read the JSON data
    print("loading JSON... ", end="", flush=True)
    with open(filename, "rt") as f:
        records = json.load(f)
    data = collections.defaultdict(list)
    for rec in records:
        rtype = rec["model"].split(".")[-1]
        datum = rec["fields"]
        if "pk" in rec:
            datum["id"] = rec["pk"]
        # convert datetime
        if rtype == "review":
            datum["creation_time"] = datetime.datetime.fromisoformat(
                datum["creation_time"]
            )

        data[rtype].append(datum)
    print("done")

    with Session() as db:

        # bulk create all the users
        bulk_insert(db, "users", data["user"], m.User)

        # bulk create all the people
        bulk_insert(db, "people", data["person"], m.Person)

        # bulk create all the movies
        bulk_insert(db, "movies", data["movie"], m.Movie)

        # bulk create all the reviews
        bulk_insert(db, "reviews", data["review"], m.Review)

        # bulk create all the directors
        bulk_insert(db, "directors", data["directors"], m.Directors)

        # bulk create all the cast
        bulk_insert(db, "cast", data["cast"], m.Cast)

        # reconcile the autoincrementing indexes with the actual indexes
        reset_sequence(db, sa.sql.quoted_name("cast", True))
        reset_sequence(db, "directors")
        reset_sequence(db, "movie")
        reset_sequence(db, "person")
        reset_sequence(db, "user")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Load a specific fixture, old data will be purged."
    )
    parser.add_argument("filename", type=str, help="The JSON dataset file")

    args = parser.parse_args()

    engine = sa.create_engine(
        os.environ.get(
            "SQLA_DSN",
            "postgresql+asyncpg://sqlalch_bench:edgedbbenchmark@"
            "localhost:15432/sqlalch_bench?async_fallback=true",
        )
    )

    load_data(args.filename, engine)

File: _sequelize/loaddata.js
"use strict";

const Sequelize = require('sequelize');
const {App} = require("./models.js");

const fs = require('fs');
const _ = require('lodash');
const ProgressBar = require('progress');

const BATCH_SIZE = 1000;


async function bulk_insert(app, data, modelName) {
  let dataName = modelName.toLowerCase();
  let adj = true;
  var bar = new ProgressBar(
    _.padEnd(modelName, 31, ' ') + '|:bar| :current/:total\t:elapsed sec',
    {total: data[dataName].length, width: 32}
  );
  bar.tick();

  for (var chunk of _.chunk(data[dataName], BATCH_SIZE)) {
    await app.transaction(async function (Model, chunk, t) {
      await Model.bulkCreate(chunk);
    }.bind(null, app.models[modelName], chunk));

    if (adj) {
      bar.tick(chunk.length - 1);
      adj = false;
    } else {
      bar.tick(chunk.length);
    }
  }
}


async function main() {
  if (process.argv.length != 3) {
    throw new Error(
      'Correct usage: node _sequelize/loaddata.js dataset.json');
  }

  const dataset_path = process.argv[2];
  var dataset = JSON.parse(fs.readFileSync(dataset_path, 'utf8'));
  var data = {};

  for (let rec of dataset) {
    let rtype = rec.model.split('.')[1];
    let datum = rec.fields;
    if (rec.pk !== undefined) {
      datum.id = rec.pk;
    }
    // convert datetime
    if (rtype == 'review') {
      datum.creation_time = new Date(datum.creation_time);
    }

    if (!(rtype in data)) {
      data[rtype] = [];
    }
    data[rtype].push(datum);
  }

  var sequelize = new App({
    pool: {
      max: 10
    }
  });
  await sequelize.sync({force: true});

  await bulk_insert(sequelize, data, 'User');
  await bulk_insert(sequelize, data, 'Person');
  await bulk_insert(sequelize, data, 'Movie');
  await bulk_insert(sequelize, data, 'Review');
  await bulk_insert(sequelize, data, 'Directors');
  await bulk_insert(sequelize, data, 'Cast');

  // creating these particular indexes is awkward in the model description
  await sequelize.getQueryInterface().addIndex('Review', {
    using: 'btree',
    fields: ['author_id'],
  });
  await sequelize.getQueryInterface().addIndex('Review', {
    using: 'btree',
    fields: ['movie_id'],
  });

  console.log('Models created.');
}

main();
